tinylog-impl/src/main/java/org/tinylog/writers/JsonWriter.java

```diff
@@ -212,71 +224,88 @@ private boolean isWhitespace(final byte character) {
 		return character == '\n' || character == '\r' || character == ' ';
 	}
 
+	/**
+	 * Removes the last occurrence of given character beginning from the end, if
+	 * it's available. Searches only for length of {@link #BUFFER_SIZE}
+	 * 
+	 * @param character Character to remove
+	 * @return true if the character was removed
+	 * @throws IOException File not found or couldn't access file
+	 */
+	private boolean removeLastOccurenceOfIfAvailable(final char character) throws IOException {
+		long sizeToTruncate = 0;
+		boolean foundChar = false;
+
+		byte[] bytes = new byte[BUFFER_SIZE];
+		randomAccessFile.seek(Math.max(0, randomAccessFile.length() - BUFFER_SIZE));
+		int numberOfBytes = randomAccessFile.read(bytes);
+
+		for (int i = numberOfBytes - 1; i >= 0; i--) {
+			byte letter = bytes[i];
+			sizeToTruncate += 1;
+
+			if (letter == character) {
```

Comment:
I think we will have a problem here with some charsets like UTF-16, which store even plain ASCII characters as two or more bytes,

Reply:
Hmm I've got a question there: Is the whole json string encoded as utf16, including the syntax like 
`--> \"test\" <--:\"asdf\"`, 
or only the content of a property e.g.
`\"test\" :\"-->asdf<--\"`

Reply:
The issue will only occur when reading data from the file as bytes. It will not occur when dealing with strings. The writer uses the same charset for the entire JSON file. With Latin based charsets like ISO 8859-1 or UTF-8, the code will work fine. However, Asians for example use often UTF-16 or other charsets that store even plain ASCII characters as two or more bytes,

Reply:
Do you think converting the byte array to String with `new String(bytes, charset)` and working with the String is a better/ the correct approach?

I got to admit that I'm a little lost :(. Sorry for troubling with such trivials.

Reply:
My approach would be to convert the wanted character (in your case `']'`) into a byte array: `byte[] characterBytes = Character.toString(']').getBytes(charset)`. This ensures that the character encoding of the JSON file is respected. Then you can search for the character bytes in the loaded byte array (`byte[] bytes = new byte[BUFFER_SIZE]`).

Reply:
Thanks for your input. I tried an attempt, but I'm afraid that's as far as I got..
```java
         private boolean removeLastOccurenceOfIfAvailable(final char character) throws IOException {
		long sizeToTruncate = 0;
		boolean foundChar = false;

		byte[] bytes = new byte[BUFFER_SIZE];
		randomAccessFile.seek(Math.max(0, randomAccessFile.length() - BUFFER_SIZE));
		int numberOfBytes = randomAccessFile.read(bytes);
		byte[] characterBytes = Character.toString(character).getBytes(charset);
		int indexOfCharacter = characterBytes.length - 1;

		for (int i = numberOfBytes - 1; i >= 0; i--) {
			byte letter = bytes[i];
			sizeToTruncate += 1;
			if (letter == characterBytes[indexOfCharacter]) {
				indexOfCharacter--;
			} else {
				indexOfCharacter = characterBytes.length - 1;
			}
			if (indexOfCharacter < 0) {
				foundChar = true;
			}
			if (foundChar && !isWhitespace(letter)) {
				break;
			}
		}
		if (!foundChar) {
			return foundChar;
		}
		long newFileSize = randomAccessFile.length() - sizeToTruncate;
		randomAccessFile.setLength(newFileSize);
		randomAccessFile.seek(randomAccessFile.length());
		return foundChar;
	}
```
which seems alright, when viewing with an editor like VSCode: 
```json
[
	{
		"message" : "Hello World!"
	}
,
	{
		"message" : "Hello World!"
	}
]
```

However, I cannot get following test to pass...
```java
       @Test
	public void handleCharset() throws IOException {
		String file = FileSystem.createTemporaryFile();
		Map<String, String> properties = new HashMap<>();
		properties.put("file", file);
		properties.put("buffered", "false");
		properties.put("append", "false");
		properties.put("charset", "UTF-16BE");
		properties.put("field.message", "message");

		JsonWriter writer;
		writer = new JsonWriter(properties);
		LogEntry givenLogEntry = LogEntryBuilder.prefilled(JsonWriterTest.class).create();
		writer.write(givenLogEntry);
		writer.close();

		byte[] expectedResult = String.format("[%s\t{%s\t\t\"message\" : \"%s\"%s\t}%s]", NEW_LINE, NEW_LINE,
				givenLogEntry.getMessage(), NEW_LINE, NEW_LINE).getBytes("UTF-16BE");
		byte[] resultingEntry = FileSystem.readFile(file).getBytes("UTF-16BE");
		assertArrayEquals(resultingEntry, expectedResult);
	}
```

Reply:
This works for unbuffured writing as long as `writer.buffered != null`:

```java
/*
 * Copyright 2021 Direnc Timur
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package org.tinylog.writers;

import java.io.File;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.charset.Charset;
import java.util.Collection;
import java.util.Collections;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Map.Entry;

import org.tinylog.Level;
import org.tinylog.core.LogEntry;
import org.tinylog.core.LogEntryValue;
import org.tinylog.pattern.FormatPatternParser;
import org.tinylog.pattern.Token;
import org.tinylog.provider.InternalLogger;
import org.tinylog.writers.raw.BufferedWriterDecorator;
import org.tinylog.writers.raw.ByteArrayWriter;
import org.tinylog.writers.raw.RandomAccessFileWriter;
import org.tinylog.writers.raw.SynchronizedWriterDecorator;

/**
 * Writer for outputting log entries to a log file in JSON format. Already
 * existing files can be continued.
 */
public final class JsonWriter implements Writer {

	private static final String NEW_LINE = System.getProperty("line.separator");
	private static final int BUFFER_SIZE = 1024;
	private static final String FIELD_PREFIX = "field.";

	private final Charset charset;
	private final RandomAccessFile randomAccessFile;
	private ByteArrayWriter writer;

	private StringBuilder builder;
	private final Map<String, Token> jsonProperties;

	private final byte[] bracketOpenBytes;
	private final byte[] bracketCloseBytes;
	private final byte[] curlyBracketBytes;
	private final byte[] spaceBytes;
	private final byte[] lineFeedBytes;
	private final byte[] carriageReturnBytes;

	/**
	 * @throws IOException              File not found or couldn't access file
	 * @throws IllegalArgumentException Log file is not defined in configuration
	 */
	public JsonWriter() throws IOException {
		this(Collections.<String, String>emptyMap());
	}

	/**
	 * @param properties Configuration for writer
	 * @throws IOException              File not found or couldn't access file
	 * @throws IllegalArgumentException Log file is not defined in configuration
	 */
	public JsonWriter(final Map<String, String> properties) throws IOException {
		jsonProperties = createTokens(properties);
		charset = getCharset(properties);

		bracketOpenBytes = "[".getBytes(charset);
		bracketCloseBytes = "]".getBytes(charset);
		curlyBracketBytes = "{".getBytes(charset);
		spaceBytes = " ".getBytes(charset);
		lineFeedBytes = "\n".getBytes(charset);
		carriageReturnBytes = "\r".getBytes(charset);

		String fileName = getFileName(properties);
		File file = new File(fileName).getAbsoluteFile();
		file.getParentFile().mkdirs();

		boolean append = Boolean.parseBoolean(properties.get("append"));

		randomAccessFile = new RandomAccessFile(file, "rw");
		writer = new RandomAccessFileWriter(randomAccessFile);

		boolean buffered = Boolean.parseBoolean(properties.get("buffered"));
		if (buffered) {
			writer = new BufferedWriterDecorator(writer);
		}

		boolean writingThread = Boolean.parseBoolean(properties.get("writingthread"));
		if (writingThread) {
			builder = new StringBuilder();
		} else {
			writer = new SynchronizedWriterDecorator(writer, randomAccessFile);
		}

		preProcessFile(append);
	}

	@Override
	public void write(final LogEntry logEntry) throws IOException {
		StringBuilder builder;
		if (this.builder == null) {
			builder = new StringBuilder();
		} else {
			builder = this.builder;
			builder.setLength(0);
		}
		removeLastOccurrenceOfIfAvailable(bracketCloseBytes);
		if (!isFirstEntry()) {
			builder.append(",");
		}
		addJsonObject(logEntry, builder);

		byte[] data = builder.toString().getBytes(charset);
		writer.write(data, data.length);
	}

	@Override
	public void flush() throws IOException {
		writer.flush();
	}

	@Override
	public void close() throws IOException {
		writer.flush();
		writer.close();
	}

	@Override
	public Collection<LogEntryValue> getRequiredLogEntryValues() {
		Collection<LogEntryValue> values = EnumSet.noneOf(LogEntryValue.class);
		for (Token token : jsonProperties.values()) {
			values.addAll(token.getRequiredLogEntryValues());
		}
		return values;
	}

	private boolean isFirstEntry() throws IOException {
		if (randomAccessFile.length() == 0) {
			return true;
		} else {
			byte[] bytes = new byte[BUFFER_SIZE];
			randomAccessFile.seek(0);
			int numberOfBytes = randomAccessFile.read(bytes);
			for (int i = numberOfBytes - 1; i >= 0; i--) {
				if (isCharacter(curlyBracketBytes, bytes, i)) {
					return false;
				}
			}
		}
		return true;
	}

	/**
	 * Prepares and adds a Json Object. Also escapes special characters.
	 *
	 * @param logEntry LogEntry with information for token
	 * @param builder  Target for the created the JSON object
	 */
	private void addJsonObject(final LogEntry logEntry, final StringBuilder builder) {
		builder.append(NEW_LINE);
		builder.append('\t');
		builder.append("{").append(NEW_LINE);

		Token[] tokenEntries = jsonProperties.values().toArray(new Token[0]);
		String[] fields = jsonProperties.keySet().toArray(new String[0]);

		for (int i = 0; i < tokenEntries.length; i++) {
			builder.append("\t\t\"").append(fields[i]).append("\" : \"");
			int start = builder.length();

			Token token = tokenEntries[i];
			token.render(logEntry, builder);

			escapeCharacter("\\", "\\\\", builder, start);
			escapeCharacter("\"", "\\\"", builder, start);
			escapeCharacter(NEW_LINE, "\\n", builder, start);
			escapeCharacter("\t", "\\t", builder, start);
			escapeCharacter("\b", "\\b", builder, start);
			escapeCharacter("\f", "\\f", builder, start);
			escapeCharacter("\n", "\\n", builder, start);
			escapeCharacter("\r", "\\r", builder, start);

			builder.append("\"");

			if (i + 1 < jsonProperties.size()) {
				builder.append(",").append(NEW_LINE);
			}
		}
		builder.append(NEW_LINE).append("\t}").append(NEW_LINE).append("]");
	}

	private void escapeCharacter(final String character, final String escapeWith, final StringBuilder stringBuilder,
								 final int startIndex) {
		for (int index = stringBuilder.indexOf(character, startIndex); index != -1; index = stringBuilder
				.indexOf(character, index + escapeWith.length())) {
			stringBuilder.replace(index, index + character.length(), escapeWith);
		}
	}

	private Charset getCharset(final Map<String, String> properties) {
		String charsetName = properties.get("charset");
		try {
			return charsetName == null ? Charset.defaultCharset() : Charset.forName(charsetName);
		} catch (IllegalArgumentException ex) {
			InternalLogger.log(Level.ERROR, "Invalid charset: " + charsetName);
			return Charset.defaultCharset();
		}
	}

	private String getFileName(final Map<String, String> properties) {
		String fileName = properties.get("file");
		if (fileName == null) {
			throw new IllegalArgumentException("File name is missing for file writer");
		}
		return fileName;
	}

	/**
	 * Removes the last occurrence of given character beginning from the end, if
	 * it's available. Searches only for length of {@link #BUFFER_SIZE}
	 *
	 * @param character Character to remove
	 * @return true if the character was removed
	 * @throws IOException File not found or couldn't access file
	 */
	private boolean removeLastOccurrenceOfIfAvailable(final byte[] character) throws IOException {
		long sizeToTruncate = 0;
		boolean foundChar = false;

		byte[] bytes = new byte[BUFFER_SIZE];
		randomAccessFile.seek(Math.max(0, randomAccessFile.length() - BUFFER_SIZE));

		for (int i = randomAccessFile.read(bytes) - character.length; i >= 0; i--) {
			sizeToTruncate += 1;

			if (isCharacter(character, bytes, i)) {
				foundChar = true;
			} else if (foundChar && !isWhitespace(bytes, i)) {
				sizeToTruncate -= 1;
				break;
			}
		}
		if (!foundChar) {
			return foundChar;
		}
		long newFileSize = randomAccessFile.length() - sizeToTruncate;
		randomAccessFile.setLength(newFileSize);
		randomAccessFile.seek(randomAccessFile.length());
		return foundChar;
	}

	/**
	 * The method checks for availability of a character in the RandomAccessFile.
	 * Only checks for length of the {@link #BUFFER_SIZE}.
	 *
	 * @param character Character to check for availability
	 * @param start     Offset where the search should begin
	 * @return boolean if the character was found
	 * @throws IOException File not found or couldn't access file
	 */
	private boolean isCharacterAvailable(final byte[] character, final long start) throws IOException {
		byte[] bytes = new byte[BUFFER_SIZE];
		randomAccessFile.seek(start);
		int numberOfBytes = randomAccessFile.read(bytes);

		for (int i = numberOfBytes - 1; i >= 0; i--) {
			if (isCharacter(character, bytes, i)) {
				return true;
			}
		}
		return false;
	}

	/**
	 * Pre-processes the JSON file. If append mode is on, deletes the closing
	 * bracket and adds a comma instead. If it's a new file, appends an opening
	 * bracket.
	 *
	 * @param append Append Mode on or off
	 * @throws IOException              Error reading or writing file
	 * @throws IllegalArgumentException Invalid file format
	 */
	private void preProcessFile(final boolean append) throws IOException, IllegalArgumentException {
		if (append && randomAccessFile.length() > 0) {
			if (!isCharacterAvailable(bracketOpenBytes, 0)
					|| !isCharacterAvailable(bracketCloseBytes, Math.max(0, randomAccessFile.length() - BUFFER_SIZE))) {
				throw new IllegalArgumentException(
						"Invalid JSON file. The file is missing either an opening or a closing bracket for the array.");
			}
		} else {
			randomAccessFile.setLength(0);
			writer.write(bracketOpenBytes, bracketOpenBytes.length);
			writer.write(bracketCloseBytes, bracketCloseBytes.length);
		}
	}

	private boolean isWhitespace(final byte[] data, final int position) {
		return isCharacter(spaceBytes, data, position)
				|| isCharacter(lineFeedBytes, data, position)
				|| isCharacter(carriageReturnBytes, data, position);
	}

	private boolean isCharacter(final byte[] character, final byte[] data, final int position) {
		if (data.length - position < character.length) {
			return false;
		} else {
			for (int i = 0; i < character.length; ++i) {
				if (character[i] != data[position + i]) {
					return false;
				}
			}
			return true;
		}
	}

	private static Map<String, Token> createTokens(final Map<String, String> properties) {
		FormatPatternParser parser = new FormatPatternParser(properties.get("exception"));

		Map<String, Token> tokens = new HashMap<String, Token>();
		for (Entry<String, String> entry : properties.entrySet()) {
			if (entry.getKey().toLowerCase(Locale.ROOT).startsWith(FIELD_PREFIX)) {
				tokens.put(entry.getKey().substring(FIELD_PREFIX.length()), parser.parse(entry.getValue()));
			}
		}
		return tokens;
	}

}
```

Reply:
I found a general issue with encoding headers like BOMs für UTF-16 (see #207). I'm going to fix that.

Reply:
Thanks for your valuable input!

I had still one issue inside `removeLastOccurenceOfIfAvailable`. We need to remove all bytes from the character to be searched, instead of just one (or possibly two)
```java
if (isCharacter(character, bytes, i)) {
    sizeToTruncate += character.length;
    foundChar = true;
} else if (foundChar) {
    break;
} else {
    sizeToTruncate += 1;
}
```

Other than that - yes, buffer mode doesn't work with that..

Reply:
I think it will be very hard to modify the contents of the buffer. I can think of 2 ways to work around it: 
1. Flush writer before attempting to write new data (which makes buffering somewhat obsolete)
2. Hold the buffer explicitly in our own BufferedWriter from where we can get and re-set it's content.

Reply:
I have thought much about solving this problem. My favorite solution is to add new methods to the `ByteArrayWriter` for reading data from the head or tail and for shrinking the file size. In this case, the `ByteArrayWriter` will provide all functionality you need for the JSON writer. You won't need anymore to use the `RandomAccessFile` directly as you can do all I/O with the `ByteArrayWriter` and you won't need to care if the output is buffered or not. I think this abstraction will simplify the JSON writer a lot.

During the week, I will make all required changes for `ByteArrayWriter` and its implementations. Afterwards you can merge the changes.

Reply:
This sounds like a good idea. I was thinking in a similar direction. If I can help you with the implementation, please feel free to involve me:).

Reply:
I have applied all the changes. I hope the new API is self-explanatory. However, I assume that you will need some effort to merge the changes properly.

Reply:
@proohit Are you getting along with the new `ByteAraryWriter` API?

Reply:
First of all, thanks for your changes. I roughly skimmed them, but haven't had a chance to inspect them in detail. Currently, I'm a little blocked by my thesis, but I will continue as soon as I have a little more buffer (pun intended :D).

Reply:
Thesis comes first! Good success :)